src/EventStore.Core/Index/TableIndex.cs

```diff
@@ -222,39 +226,49 @@ public void AddEntries(long commitPos, IList<IndexKey> entries)
                 {
                     prepareCheckpoint = Math.Max(prepareCheckpoint, collection[i].Position);
                 }
+                
+                TryProcessAwaitingTables(commitPos, prepareCheckpoint, _autoMergeIndexes);
+            }
+        }
 
-                lock (_awaitingTablesLock)
-                {
-                    var newTables = new List<TableItem> { new TableItem(_memTableFactory(), -1, -1) };
-                    newTables.AddRange(_awaitingMemTables.Select(
-                        (x, i) => i == 0 ? new TableItem(x.Table, prepareCheckpoint, commitPos) : x));
+        public void MergeIndexes()
+        {
+            TryProcessAwaitingTables(_indexMap.CommitCheckpoint, _indexMap.PrepareCheckpoint, true);
```

Comment:
It is not correct to call `TryProcessAwaitingTables(long commitPos, long prepareCheckpoint, bool mergeIndexes)` here. The function should be called only when a MemTable is completed.

**Fix:**
`TryProcessAwaitingTables(bool mergeIndexes)` should directly be called.

**Issues:**
1. It will create a new Memtable everytime even if the previous one has not been completed. (e.g. it will even dump a MemTable with 0 index entries to a PTable)
2. The supplied values: `_indexMap.CommitCheckpoint` and `_indexMap.PrepareCheckpoint` are not correct and causes a crash when calling `/admin/mergeindexes` on an empty DB:

```
[21019,08,06:41:13.795] Error in TableIndex.ReadOffQueue
EXCEPTION OCCURRED
prepareCheckpoint should be non negative.
Parameter name: prepareCheckpoint
[21019,08,06:41:13.799] Global Unhandled Exception occurred.
EXCEPTION OCCURRED
prepareCheckpoint should be non negative.
Parameter name: prepareCheckpoint
[ERROR] FATAL UNHANDLED EXCEPTION: System.ArgumentOutOfRangeException: prepareCheckpoint should be non negative.
Parameter name: prepareCheckpoint
  at EventStore.Common.Utils.Ensure.Nonnegative (System.Int64 number, System.String argumentName) [0x00011] in <82749b1b247d4718bba6fcc8de345a00>:0 
  at EventStore.Core.Index.IndexMap.AddPTable (EventStore.Core.Index.PTable tableToAdd, System.Int64 prepareCheckpoint, System.Int64 commitCheckpoint, System.Func`3[T1,T2,TResult] upgradeHash, System.Func`2[T,TResult] existsAt, System.Func`2[T,TResult] recordExistsAt, EventStore.Core.Index.IIndexFilenameProvider filenameProvider, System.Byte version, System.Int32 indexCacheDepth, System.Boolean skipIndexVerify, System.Boolean mergeIndexes) [0x00000] in <3aba2e5e6d364a458c766d7dd95c2f8f>:0 
  at EventStore.Core.Index.TableIndex.ReadOffQueue (System.Boolean mergeIndexes) [0x00288] in <3aba2e5e6d364a458c766d7dd95c2f8f>:0 
  at EventStore.Core.Index.TableIndex+<>c__DisplayClass41_0.<TryProcessAwaitingTables>b__0 (System.Object x) [0x00000] in <3aba2e5e6d364a458c766d7dd95c2f8f>:0 
  at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context (System.Object state) <0x7ff388e56140 + 0x0002a> in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) <0x7ff388e50ee0 + 0x001a2> in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () <0x7ff388e560f0 + 0x00025> in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00074] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () <0x7ff388e55f70 + 0x00005> in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
```

Reply:
Also:
1. If a background index merge/scavenge task is already running, the merge will not take place and user will not be informed. A check can be done on `_backgroundRunning = True` to log something and return. (this is particularly possible if you call `admin/mergeindexes` multiple times within a short amount of time)

2. Improvement: if `ManualIndexMerge` is set to `False`, the functionality can be disabled and the function can just log something and return (to be on the safe side for users who have not enabled this functionality)

Reply:
Manually trigger the merge operation requires to pass through the 
TryProcessAwaitingTables(long commitPos, long prepareCheckpoint, bool mergeIndexes) method. I added a filter to avoid the exception.
1. Now it's fixed. If you send multiple http commands you'll get a log saying that the operation is running.
2. Even when the AutoMergeIndexes is set to True (normal behaviour) it is possible to manually trigger the Index Merge operation. This is a feature and not having it is not an improvement.